File: project/app/src/main/java/org/github/henryquan/autoclicker/FloatingManager.java
package org.github.henryquan.autoclicker;

import android.content.Context;
import android.view.View;
import android.view.WindowManager;

public class FloatingManager {
    private static FloatingManager mInstance;
    private WindowManager mWindowManager;
    private Context mContext;

    private FloatingManager(Context context) {
        mContext = context;
        mWindowManager = (WindowManager) mContext.getSystemService(
                Context.WINDOW_SERVICE); //获得WindowManager对象
    }

    public static FloatingManager getInstance(Context context) {
        if (mInstance == null) {
            mInstance = new FloatingManager(context);
        }
        return mInstance;
    }

    /**
     * adds floating view
     *
     * @param view
     * @param params
     * @return
     */
    protected boolean addView(View view, WindowManager.LayoutParams params) {
        try {
            mWindowManager.addView(view, params);
            return true;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }

    /**
     * removes floating view
     *
     * @param view
     * @return
     */
    protected boolean removeView(View view) {
        try {
            mWindowManager.removeView(view);
            return true;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }

    /**
     * @param view
     * @param params
     * @return
     */
    protected boolean updateView(View view, WindowManager.LayoutParams params) {
        try {
            mWindowManager.updateViewLayout(view, params);
            return true;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }
}

File: project/app/src/main/java/org/github/henryquan/autoclicker/FloatingManager.kt
package org.github.henryquan.autoclicker

import android.content.Context
import android.view.View
import android.view.WindowManager

class FloatingManager private constructor(private val mContext: Context) {
    private val mWindowManager: WindowManager

    /**
     * adds floating view
     *
     * @param view
     * @param params
     * @return
     */
    protected fun addView(view: View?, params: WindowManager.LayoutParams?): Boolean {
        try {
            mWindowManager.addView(view, params)
            return true
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return false
    }

    /**
     * removes floating view
     *
     * @param view
     * @return
     */
    protected fun removeView(view: View?): Boolean {
        try {
            mWindowManager.removeView(view)
            return true
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return false
    }

    /**
     * @param view
     * @param params
     * @return
     */
    protected fun updateView(view: View?, params: WindowManager.LayoutParams?): Boolean {
        try {
            mWindowManager.updateViewLayout(view, params)
            return true
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return false
    }

    companion object {
        private var mInstance: FloatingManager? = null
        fun getInstance(context: Context): FloatingManager? {
            if (mInstance == null) {
                mInstance = FloatingManager(context)
            }
            return mInstance
        }
    }

    init {
        mWindowManager = mContext.getSystemService(
            Context.WINDOW_SERVICE
        ) as WindowManager //获得WindowManager对象
    }
}